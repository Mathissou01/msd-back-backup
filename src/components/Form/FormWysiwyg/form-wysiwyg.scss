@use "src/styles/01-settings/settings.colors" as colors;
@use "src/styles/02-tools/tools.inputs" as inputs;

.c-FormWysiwyg {
  display: flex;
  flex-direction: column;

  &__Input {
    position: relative;
    display: flex;
    flex-direction: column;
    width: 100%;

    &:empty {
      background: colors.$gray-100;
      border: 1px solid colors.$blue-3;
      border-radius: 10px;
    }

    &_invalid {
      border-radius: 10px;
      outline: 1px solid colors.$error;

      &::after {
        position: absolute;
        top: 50%;
        right: 44px;
        width: 20px;
        height: 20px;
        background-image: url("data:image/svg+xml,%3Csvg width='36' height='36' viewBox='9 9 18 18' fill='none' xmlns='http://www.w3.org/2000/svg'%3E%3Cg clip-path='url(%23a)' stroke='%23EB4A65' stroke-width='1.5' stroke-linecap='round'%3E%3Cpath d='M18 22.5a.25.25 0 0 0-.049.495.25.25 0 0 0 .299-.245.25.25 0 0 0-.25-.25Z' stroke-linejoin='round'/%3E%3Cpath d='M18 20.5v-5'/%3E%3Cpath d='M19.08 11.173a1.204 1.204 0 0 0-2.16 0L10.603 24.04a1.013 1.013 0 0 0 .91 1.461h12.971a1.014 1.014 0 0 0 .911-1.461L19.08 11.174Z' stroke-linejoin='round'/%3E%3C/g%3E%3Cdefs%3E%3CclipPath id='a'%3E%3Cpath fill='%23fff' d='M9 9h18v18H9z'/%3E%3C/clipPath%3E%3C/defs%3E%3C/svg%3E");
        background-repeat: no-repeat;
        background-position: center;
        background-size: 20px;
        content: "";
        pointer-events: none;
      }

      &:focus-visible {
        outline-color: colors.$error;
      }
    }

    &_disabled {
      &::before {
        position: absolute;
        z-index: 1;
        width: 100%;
        height: 100%;
        background: colors.$gray-200;
        border: 1px solid colors.$blue-4;
        border-radius: 10px;
        content: "";
        pointer-events: none;
      }
    }

    textarea {
      display: none;
    }
  }
}
