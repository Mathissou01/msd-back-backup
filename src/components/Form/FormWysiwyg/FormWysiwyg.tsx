import _ from "lodash";
import classNames from "classnames";
import { ErrorMessage } from "@hookform/error-message";
import { Controller, useFormContext } from "react-hook-form";
import { Editor as TinyMceEditor } from "tinymce";
import React, { useRef } from "react";
import CommonErrorText from "../../Common/CommonErrorText/CommonErrorText";
import FormLabel from "../FormLabel/FormLabel";
import WysiwygEditor from "./WysiwygEditor/WysiwygEditor";
import "./form-wysiwyg.scss";

interface IFormInputProps {
  name: string;
  label: string;
  isVisible: boolean;
  validationLabel?: string;
  isRequired?: boolean;
  isDisabled?: boolean;
  defaultValue?: string;
}

export default function FormWysiwyg({
  name,
  label,
  isVisible,
  validationLabel,
  isRequired = false,
  isDisabled = false,
  defaultValue,
}: IFormInputProps) {
  /* Static Data */
  const errorMessages = {
    required: "Ce champ est obligatoire",
  };

  /* Local Data */
  const {
    control,
    formState: { isSubmitting, errors },
  } = useFormContext();
  const editorRef = useRef<TinyMceEditor | null>(null);

  return (
    <div className="c-FormWysiwyg">
      <FormLabel
        forId={name}
        label={label}
        isRequired={isRequired}
        validationLabel={validationLabel}
      >
        <Controller
          control={control}
          name={name}
          rules={{
            required: { value: isRequired, message: errorMessages.required },
          }}
          defaultValue={defaultValue}
          render={({ field: { onChange, value, ref } }) => {
            return (
              <div
                className={classNames("c-FormWysiwyg__Input", {
                  "c-FormWysiwyg__Input_invalid": _.get(errors, name),
                })}
                ref={ref}
                id={name}
                data-testid="form-wysiwyg"
              >
                {isVisible && (
                  <WysiwygEditor
                    id={name}
                    forwardedRef={editorRef}
                    onEditorChange={onChange}
                    value={value}
                    isDisabled={isSubmitting || isDisabled}
                  />
                )}
              </div>
            );
          }}
        />
      </FormLabel>
      <ErrorMessage
        errors={errors}
        name={name}
        render={({ message }) => (
          <CommonErrorText message={message} errorId={`${name}_error`} />
        )}
      />
    </div>
  );
}
