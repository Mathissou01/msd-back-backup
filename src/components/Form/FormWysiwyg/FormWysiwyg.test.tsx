import { render, screen } from "@testing-library/react";
import { FormProvider, useForm } from "react-hook-form";
import React, { ReactNode } from "react";
import FormWysiwyg from "./FormWysiwyg";

const mock = {
  name: "form-wysiwyg",
  label: "form wysiwyg",
  validationLabel: "validation label",
};

const Wrapper = (props: { children: ReactNode }) => {
  const formMethods = useForm();
  return (
    <>
      <FormProvider {...formMethods}>{props.children}</FormProvider>
    </>
  );
};

describe("FormWysiwyg", () => {
  it("renders", async () => {
    const { container } = render(
      <Wrapper>
        <FormWysiwyg
          name={mock.name}
          label={mock.label}
          validationLabel={mock.validationLabel}
          isVisible
          isRequired
        />
      </Wrapper>,
    );

    const input = await screen.findByTestId("form-wysiwyg");
    expect(input).not.toHaveClass("c-FormWysiwyg__Input_invalid");
    expect(container).toMatchSnapshot();
  });
});
