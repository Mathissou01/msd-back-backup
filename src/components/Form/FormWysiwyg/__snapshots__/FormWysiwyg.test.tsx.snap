// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormWysiwyg renders 1`] = `
<div>
  <div
    class="c-FormWysiwyg"
  >
    <div
      class="c-FormLabel"
    >
      <label
        class="c-FormLabel__LabelWrapper"
        for="form-wysiwyg"
      >
        <span
          class="c-FormLabel__Label"
        >
          form wysiwyg *
        </span>
        <span
          class="c-FormLabel__Validation"
        >
          validation label
        </span>
      </label>
      <div
        class="c-FormLabel__Content"
      >
        <div
          class="c-FormWysiwyg__Input"
          data-testid="form-wysiwyg"
          id="form-wysiwyg"
        >
          <textarea
            id="wysiwyg-editor-form-wysiwyg"
            style="visibility: hidden;"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
